extensions/amp-skimlinks/0.1/amp-skimlinks.js

```diff
@@ -0,0 +1,239 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {CommonSignals} from '../../../src/common-signals';
+import {Services} from '../../../src/services';
+import {once} from '../../../src/utils/function';
+import {whenDocumentReady} from '../../../src/document-ready';
+
+import {Tracking} from './tracking';
+
+import {AffiliateLinkResolver} from './affiliate-link-resolver';
+import {SKIMLINKS_REWRITER_ID} from './constants';
+import {EVENTS as linkRewriterEvents} from './link-rewriter/constants';
+
+import {LinkRewriterManager} from './link-rewriter/link-rewriter-manager';
+import {Waypoint} from './waypoint';
+import {getAmpSkimlinksOptions} from './skim-options';
+import {getBoundFunction} from './utils';
+
+export class AmpSkimlinks extends AMP.BaseElement {
+  /**
+   * @param {!AmpElement} element
+   */
+  constructor(element) {
+    super(element);
+
+    /** @private {?../../../src/service/xhr-impl.Xhr} */
+    this.xhr_ = null;
+
+    /** @private {?../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampDoc_ = null;
+
+    /** @private {?../../../src/service/document-info-impl.DocumentInfoDef} */
+    this.docInfo_ = null;
+
+    /** @private {?../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = null;
+
+    /** @private {?./link-rewriter/link-rewriter-manager.LinkRewriterManager} */
+    this.linkRewriterService_ = null;
+
+    /** @private {?Object} */
+    this.skimOptions_ = null;
+
+    /** @private {?./tracking.Tracking} */
+    this.trackingService_ = null;
+
+    /** @private {?./affiliate-link-resolver.AffiliateLinkResolver} */
+    this.affiliateLinkResolver_ = null;
+
+    /** @private {?./waypoint.Waypoint} */
+    this.waypoint_ = null;
+
+    /** @private {?./link-rewriter/link-rewriter.LinkRewriter} */
+    this.skimlinksLinkRewriter_ = null;
+
+    /** @private {?string} */
+    this.referrer_ = null;
+  }
+
+  /** @override */
+  buildCallback() {
```

Comment:
let's do all those in `layoutCallback()`.

scanning through the doc in buildCallback is a too heavy work

Reply:
I am happy to move it to `layoutCallback` but I thought this was only for extension with a UI (which is not the case for skimlinks).
We were thinking of using the `nodisplay` layout (`<amp-skimlinks layout="nodisplay">` ) since there is nothing to render, but this layout does not call the `layoutCallback` callback. Should we use another type of layout? Which one is the most suitable since we don't have any UI to render?

Reply:
@choumx I checked what chunk.js does and it seems like something that could be applied here if performance was an issue. However, I have also spent some time analyzing the performance of amp-skimlinks and more specifically the DOM scanning part. My analysis pointed out an issue which was making the DOM scanning slower than it should and the issue is now fixed. See https://github.com/ampproject/amphtml/pull/17907/commits/b96f6b97193a78a6ff52c5c8ada84567bc6e28c8

My new results make me think that using chunk.js would be an over-optimization considering that amp-skimlinks is not making any heavy work and the scan is running relatively quickly (less than 7ms on a real page with 168 links). The scan relies mostly on querySelectorAll which seems fast enough and runs under 1ms. I have also identified one other possible micro-optimization to bring down the processing time even more. (I would bring this change later in a different PR if needed). 

I would like to avoid using chunk.js since it seems to delay the scan of the page by at least 1 second and amp-skimlinks may miss some clicks & page impressions. In my opinion, amp-skimlinks does not bring any slowdown but if you still think it does, I think we should try to evaluate and measure it first and see if we can fix it without chunk.js. 

What specific function(s) in this PR do you consider as to heavy in terms of processing? And what's the maximum amount of time the buildCallback should take?

Reply:
Thanks for investigating this. What device did you instrument for the 7ms and 168 link figure? 

Part of AMP's [design principles](https://www.ampproject.org/about/amp-design-principles/) is to aggressively prioritize user content over analytics/ads/etc. This is especially important for ensuring good performance on low-end devices with slow network connections. 

I think that choosing O(seconds) delay in operation of amp-skimlinks rather than resource contention with page content is the right tradeoff here. Arguably such quick bounces do not have a strong user intent signal anyways.

/cc @lannka

Reply:
Ok, I have implemented the chunk solution here https://github.com/skimhub/amphtml/pull/13. Is this what you had in mind?

Reply:
Yep, thanks!